File: backend/crates/pm-ws/src/circuit_breaker.rs
use std::sync::RwLock;
use std::sync::atomic::{AtomicU32, AtomicU64, Ordering};
use std::time::{Duration, Instant};

// Conservative defaults for database circuit breaker:
// - Open after 5 failures in 60 seconds (failure_threshold/failure_window)
// - Stay open for 30 seconds before testing recovery (open_duration)
// - Require 3 consecutive successes to fully close (half_open_success_threshold)
const DEFAULT_FAILURE_THRESHOLD: u32 = 5;
const DEFAULT_OPEN_DURATION_SECS: u64 = 30;
const DEFAULT_HALF_OPEN_SUCCESS_THRESHOLD: u32 = 3;
const DEFAULT_FAILURE_WINDOW_SECS: u64 = 60;

/// Circuit breaker states
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum CircuitState {
    /// Normal operation - requests flow through
    Closed,
    /// Too many failures - requests rejected immediately
    Open,
    /// Testing if service recovered - limited requests allowed
    HalfOpen,
}

/// Circuit breaker configuration
#[derive(Debug, Clone)]
pub struct CircuitBreakerConfig {
    /// Number of failures before opening circuit
    pub failure_threshold: u32,
    /// Duration to keep circuit open before testing
    pub open_duration: Duration,
    /// Number of successful requests in half-open to close circuit
    pub half_open_success_threshold: u32,
    /// Window for counting failures
    pub failure_window: Duration,
}

impl Default for CircuitBreakerConfig {
    fn default() -> Self {
        Self {
            failure_threshold: DEFAULT_FAILURE_THRESHOLD,
            open_duration: Duration::from_secs(DEFAULT_OPEN_DURATION_SECS),
            half_open_success_threshold: DEFAULT_HALF_OPEN_SUCCESS_THRESHOLD,
            failure_window: Duration::from_secs(DEFAULT_FAILURE_WINDOW_SECS),
        }
    }
}

/// Thread-safe circuit breaker
pub struct CircuitBreaker {
    config: CircuitBreakerConfig,
    state: RwLock<CircuitState>,
    failure_count: AtomicU32,
    success_count: AtomicU32,
    last_failure_time: AtomicU64,
    opened_at: AtomicU64,
}

impl CircuitBreaker {
    pub fn new(config: CircuitBreakerConfig) -> Self {
        Self {
            config,
            state: RwLock::new(CircuitState::Closed),
            failure_count: AtomicU32::new(0),
            success_count: AtomicU32::new(0),
            last_failure_time: AtomicU64::new(0),
            opened_at: AtomicU64::new(0),
        }
    }

    /// Check if request should be allowed
    pub fn allow_request(&self) -> Result<(), CircuitBreakerError> {
        let state = *self.state.read().unwrap();

        match state {
            CircuitState::Closed => Ok(()),
            CircuitState::Open => {
                // Check if we should transition to half-open
                let opened_at = self.opened_at.load(Ordering::SeqCst);
                let now = Instant::now().elapsed().as_secs();

                if now - opened_at >= self.config.open_duration.as_secs() {
                    // Transition to half-open
                    let mut state_guard = self.state.write().unwrap();
                    if *state_guard == CircuitState::Open {
                        *state_guard = CircuitState::HalfOpen;
                        self.success_count.store(0, Ordering::SeqCst);
                        log::info!("Circuit breaker transitioning to HalfOpen");
                    }
                    Ok(())
                } else {
                    Err(CircuitBreakerError::CircuitOpen {
                        retry_after_secs: self.config.open_duration.as_secs() - (now - opened_at),
                    })
                }
            }
            CircuitState::HalfOpen => Ok(()), // Allow limited requests in half-open
        }
    }

    /// Record a successful request
    pub fn record_success(&self) {
        let state = *self.state.read().unwrap();

        match state {
            CircuitState::Closed => {
                // Reset failure count on success
                self.failure_count.store(0, Ordering::SeqCst);
            }
            CircuitState::HalfOpen => {
                let successes = self.success_count.fetch_add(1, Ordering::SeqCst) + 1;
                if successes >= self.config.half_open_success_threshold {
                    // Close the circuit
                    let mut state_guard = self.state.write().unwrap();
                    *state_guard = CircuitState::Closed;
                    self.failure_count.store(0, Ordering::SeqCst);
                    log::info!("Circuit breaker closed after {} successes", successes);
                }
            }
            CircuitState::Open => {} // Shouldn't happen, but ignore
        }
    }

    /// Record a failed request
    pub fn record_failure(&self) {
        let now = std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap()
            .as_secs();

        let last_failure = self.last_failure_time.load(Ordering::SeqCst);

        // Reset count if outside failure window
        if now - last_failure > self.config.failure_window.as_secs() {
            self.failure_count.store(0, Ordering::SeqCst);
        }

        self.last_failure_time.store(now, Ordering::SeqCst);
        let failures = self.failure_count.fetch_add(1, Ordering::SeqCst) + 1;

        let state = *self.state.read().unwrap();

        match state {
            CircuitState::Closed => {
                if failures >= self.config.failure_threshold {
                    // Open the circuit
                    let mut state_guard = self.state.write().unwrap();
                    *state_guard = CircuitState::Open;
                    self.opened_at
                        .store(Instant::now().elapsed().as_secs(), Ordering::SeqCst);
                    log::warn!("Circuit breaker OPEN after {} failures", failures);
                }
            }
            CircuitState::HalfOpen => {
                // Single failure in half-open reopens circuit
                let mut state_guard = self.state.write().unwrap();
                *state_guard = CircuitState::Open;
                self.opened_at
                    .store(Instant::now().elapsed().as_secs(), Ordering::SeqCst);
                log::warn!("Circuit breaker reopened due to failure in HalfOpen state");
            }
            CircuitState::Open => {} // Already open
        }
    }

    /// Get current state
    pub fn state(&self) -> CircuitState {
        *self.state.read().unwrap()
    }
}

#[derive(Debug, Clone)]
pub enum CircuitBreakerError {
    CircuitOpen { retry_after_secs: u64 },
}

impl std::fmt::Display for CircuitBreakerError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::CircuitOpen { retry_after_secs } => {
                write!(
                    f,
                    "Circuit breaker open. Retry after {} seconds",
                    retry_after_secs
                )
            }
        }
    }
}

impl std::error::Error for CircuitBreakerError {}

File: frontend/ProjectManagement.Services/Resilience/CircuitBreaker.cs
using System.IO;
using System.Net.WebSockets;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Options;
using ProjectManagement.Core.Exceptions;

namespace ProjectManagement.Services.Resilience;

/// <summary>
/// Circuit breaker to prevent cascading failures.
/// Matches backend pm-ws circuit breaker behavior.
/// </summary>
public sealed class CircuitBreaker
{
    private readonly CircuitBreakerOptions _options;
    private readonly ILogger<CircuitBreaker> _logger;
    private readonly object _lock = new();

    private CircuitState _state = CircuitState.Closed;
    private int _failureCount;
    private int _successCount;
    private DateTime _lastFailureTime = DateTime.MinValue;
    private DateTime _openedAt = DateTime.MinValue;

    public CircuitState State
    {
        get
        {
            lock (_lock) return _state;
        }
    }

    public TimeSpan? RetryAfter
    {
        get
        {
            lock (_lock)
            {
                if (_state != CircuitState.Open)
                    return null;

                var elapsed = DateTime.UtcNow - _openedAt;
                var remaining = _options.OpenDuration - elapsed;
                return remaining > TimeSpan.Zero ? remaining : TimeSpan.Zero;
            }
        }
    }

    public CircuitBreaker(
        IOptions<CircuitBreakerOptions> options,
        ILogger<CircuitBreaker> logger)
    {
        _options = options.Value;
        _logger = logger;
    }

    /// <summary>
    /// Check if a request should be allowed through.
    /// </summary>
    public bool AllowRequest()
    {
        lock (_lock)
        {
            switch (_state)
            {
                case CircuitState.Closed:
                    return true;

                case CircuitState.Open:
                    // Check if we should transition to half-open
                    if (DateTime.UtcNow - _openedAt >= _options.OpenDuration)
                    {
                        _state = CircuitState.HalfOpen;
                        _successCount = 0;
                        _logger.LogInformation("Circuit breaker transitioning to HalfOpen");
                        return true;
                    }

                    return false;

                case CircuitState.HalfOpen:
                    return true;

                default:
                    return false;
            }
        }
    }

    /// <summary>
    /// Record a successful operation.
    /// </summary>
    public void RecordSuccess()
    {
        lock (_lock)
        {
            switch (_state)
            {
                case CircuitState.Closed:
                    _failureCount = 0;
                    break;

                case CircuitState.HalfOpen:
                    _successCount++;
                    if (_successCount >= _options.HalfOpenSuccessThreshold)
                    {
                        _state = CircuitState.Closed;
                        _failureCount = 0;
                        _logger.LogInformation(
                            "Circuit breaker closed after {Successes} successes",
                            _successCount);
                    }

                    break;
            }
        }
    }

    /// <summary>
    /// Record a failed operation.
    /// </summary>
    public void RecordFailure()
    {
        lock (_lock)
        {
            var now = DateTime.UtcNow;

            // Reset count if outside failure window
            if (now - _lastFailureTime > _options.FailureWindow)
            {
                _failureCount = 0;
            }

            _lastFailureTime = now;
            _failureCount++;

            switch (_state)
            {
                case CircuitState.Closed:
                    if (_failureCount >= _options.FailureThreshold)
                    {
                        _state = CircuitState.Open;
                        _openedAt = now;
                        _logger.LogWarning(
                            "Circuit breaker OPEN after {Failures} failures",
                            _failureCount);
                    }

                    break;

                case CircuitState.HalfOpen:
                    _state = CircuitState.Open;
                    _openedAt = now;
                    _logger.LogWarning(
                        "Circuit breaker reopened due to failure in HalfOpen state");
                    break;
            }
        }
    }

    /// <summary>
    /// Execute an operation with circuit breaker protection.
    /// </summary>
    public async Task<T> ExecuteAsync<T>(
        Func<CancellationToken, Task<T>> operation,
        CancellationToken ct = default)
    {
        if (!AllowRequest())
        {
            throw new CircuitOpenException(RetryAfter ?? _options.OpenDuration);
        }

        try
        {
            var result = await operation(ct);
            RecordSuccess();
            return result;
        }
        catch (Exception ex)
        {
            // Record failure for transient errors, but always re-throw
            if (ShouldRecordAsFailure(ex))
            {
                RecordFailure();
            }

            throw;
        }
    }

    private static bool ShouldRecordAsFailure(Exception ex)
    {
        // Don't count validation errors or cancellation as circuit failures
        return ex is not ValidationException
               && ex is not OperationCanceledException
               && ex is not VersionConflictException;
    }
}